subgraphs/playback/src/utils/server.ts

```diff
@@ -0,0 +1,181 @@
+        (ctx.extra.request.headers?.['authorization'] as string) ??
+        '';
+      checkRouterSecret(routerAuthorization);
+      const token =
+        (ctx.connectionParams?.['authorization'] as string) ??
+        (ctx.extra.request.headers?.['authorization'] as string);
+
+      const defaultCountryCode =
+        (ctx.connectionParams?.['country-code'] as string) ??
+        (ctx.extra.request.headers?.['country-code'] as string) ??
+        'US';
+
+      if (!token) {
+        const userIdForMocks =
+          ctx.extra.request.rawHeaders['x-graphos-id'] ?? 'default';
+        addUser(userIdForMocks);
+
+        return {
+          defaultCountryCode,
+          dataSources: {
+            spotify: new MockedSpotifyDataSource(userIdForMocks),
+          },
+          userIdForMocks,
+        };
+      }
+
+      return {
+        defaultCountryCode,
+        dataSources: {
+          spotify: new SpotifyAPI({
+            cache: callbackApolloServer.cache,
+            token,
+          }),
+        },
+      };
+    },
+  },
+  wsServer
+);
+
+const sentryPlugin: ApolloServerPlugin<ContextValue> = {
+  async requestDidStart() {
+    return {
+      async didEncounterErrors(ctx) {
+        for (const err of ctx.errors) {
+          if (err.extensions?.code == 'GRAPHQL_VALIDATION_FAILED')
+            Sentry.withScope((scope) => {
+              // Annotate whether failing operation was query/mutation/subscription
+              scope.setTag('message', err.message);
+              // Log query and variables as extras
+              // (make sure to strip out sensitive data!)
+              scope.setExtra('query', ctx.request.query);
+              scope.setExtra('variables', ctx.request.variables);
+              if (err.path) {
+                // We can also add the path as breadcrumb
+                scope.addBreadcrumb({
+                  category: 'query-path',
+                  message: err.path.join(' > '),
+                  level: 'debug',
+                });
+              }
+              Sentry.captureException(err);
+            });
+          else if (!ctx.operation) return;
+          else
+            Sentry.withScope((scope) => {
+              // Annotate whether failing operation was query/mutation/subscription
+              scope.setTag('kind', ctx.operation.operation);
+              // Log query and variables as extras
+              // (make sure to strip out sensitive data!)
+              scope.setExtra('query', ctx.request.query);
+              scope.setExtra('variables', ctx.request.variables);
+              if (err.path) {
+                // We can also add the path as breadcrumb
+                scope.addBreadcrumb({
+                  category: 'query-path',
+                  message: err.path.join(' > '),
+                  level: 'debug',
+                });
+              }
+              Sentry.captureException(err);
+            });
```

Comment:
```suggestion
          else if (!ctx.operation) {
            return
          } else {
            Sentry.withScope((scope) => {
              // Annotate whether failing operation was query/mutation/subscription
              scope.setTag('kind', ctx.operation.operation);
              // Log query and variables as extras
              // (make sure to strip out sensitive data!)
              scope.setExtra('query', ctx.request.query);
              scope.setExtra('variables', ctx.request.variables);
              if (err.path) {
                // We can also add the path as breadcrumb
                scope.addBreadcrumb({
                  category: 'query-path',
                  message: err.path.join(' > '),
                  level: 'debug',
                });
              }
              Sentry.captureException(err);
            });
          }
```

This is much easier to reason about for me since I don't have to use visual indentation to try and match up what code blocks belong under what.

Reply:
Implemented in [9ac8554](https://github.com/apollographql/spotify-showcase/pull/94/commits/9ac8554784f26dee7c5f80253db7fb7953b3a44a)